text/0049-soc-gpio-peripheral.md

```diff
@@ -76,13 +76,33 @@ Each `Mode.pin_x` field can hold the following values:
 ```python3
 class Mode(enum.Enum, shape=unsigned(2)):
     INPUT_ONLY = 0b00
-    PUSH_PULL  = 0b10
-    OPEN_DRAIN = 0b01
+    PUSH_PULL  = 0b01
+    OPEN_DRAIN = 0b10
+    ALTERNATE  = 0b11
 ```
 
-- If `Mode.pin_x` is `INPUT_ONLY`, then `pins[x].oe = 0`.
-- If `Mode.pin_x` is `PUSH_PULL`, then `pins[x].oe = 1` and `pins[x].o = Output.pin_x`.
-- If `Mode.pin_x` is `OPEN_DRAIN`, then `pins[x].oe = ~Output.pin_x` and `pins[x].o = 0`.
+Each `Mode.pin_x` field resets to `INPUT_ONLY`.
+
+If `Mode.pin_x` is `INPUT_ONLY`:
+- `pins[x].alt_mode` returns 0.
+- `Input.pin_x` returns the last value of `pins[x].i` sampled on a clock cycle.
+- `pins[x].oe` returns 0 `pins[x].o` return 0.
+
+If `Mode.pin_x` is `PUSH_PULL`:
+- `pins[x].alt_mode` returns 0.
+- `Input.pin_x` returns the last value of `pins[x].i` sampled on a clock cycle.
+- `pins[x].oe` returns 1 and `pins[x].o` returns `Output.pin_x`.
+
+If `Mode.pin_x` is `OPEN_DRAIN`:
+- `pins[x].alt_mode` returns 0.
+- `Input.pin_x` returns the last value of `pins[x].i` sampled on a clock cycle.
+- `pins[x].oe` returns `~Output.pin_x` and `pins[x].o` returns 0.
+
+If `Mode.pin_x` is `ALTERNATE`:
+- `pins[x].alt_mode` returns 1.
+- `Input.pin_x`, `pins[x].oe` and `pins[x].o` return implementation-specific values.
```

Comment:
I don't think that works very well--`pins[x].oe` is controlled by the GPIO peripheral, so there's going to be one specific behavior that we implement. I suggest this behavior:
- `pins[x].oe` is 0,
- `pins[x].o` is connected to `Output.pin_x`,
- `Input.pin_x` is connected to `pins[x].i`].

However, we also note for the *users* of the peripheral that given the intended use of the `ALTERNATE` mode in the SoC, `Input` and `Output` registers have (SoC) implementation specific functions.

Reply:
"return" is an odd verb here but I think Catherine's described behavior is good.